### PR TITLE
Fix the map concurrent read/write issue in deployment controller

### LINKED ISSUE #29960: Concurrent RS labels map read map write in deployment controller
`equalIgnoreHash` may produce `fatal error: concurrent map read and map write` because the labels maps are read and write concurrently. Will fix this with the following steps:

1. Compare the labels too see if they're equal (ignore hash label)
2. Then set the template labels to `nil` and do a deep equal 

@kubernetes/deployment @lavalamp

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #29960

@lavalamp @kubernetes/deployment

### CODE REVIEW

File: pkg/controller/deployment/util/deployment_util.go [hunk 1]
```diff
@@ -402,20 +403,39 @@ func ListPods(deployment *extensions.Deployment, getPodList podListFunc) (*api.P
...
+	if len(labels1) == 0 || len(labels2) == 0 {
 		return false, fmt.Errorf("Unexpected empty labels found in given template")
 	}
-	hash1 := template1.Labels[extensions.DefaultDeploymentUniqueLabelKey]
-	hash2 := template2.Labels[extensions.DefaultDeploymentUniqueLabelKey]
-	// compare equality ignoring pod-template-hash
-	template1.Labels[extensions.DefaultDeploymentUniqueLabelKey] = hash2
+	if len(labels1) != len(labels2) {
+		// If length is different, and it's not because of a missing hash label
+		if math.Abs(float64(len(labels1)-len(labels2))) > 1 || (containsHash(labels1) && containsHash(labels2)) || (!containsHash(labels1) && !containsHash(labels2)) {
```
Comment (deads2k) [inline pkg/controller/deployment/util/deployment_util.go]: `if len(labels1) != len(labels2) && containsHash(label1) == containsHash(label2)` ?

File: pkg/controller/deployment/util/deployment_util.go [hunk 2]
```diff
@@ -402,20 +403,39 @@ func ListPods(deployment *extensions.Deployment, getPodList podListFunc) (*api.P
...
+			return false, nil
+		}
+	}
+
+	// Then, compare the templates without comparing their labels
+	template1.Labels = nil
+	template2.Labels = nil
 	result := api.Semantic.DeepEqual(template1, template2)
-	template1.Labels[extensions.DefaultDeploymentUniqueLabelKey] = hash1
+	template1.Labels = labels1
```
Comment (deads2k) [inline pkg/controller/deployment/util/deployment_util.go]: This is a copy of the podtemplatespec, no need to restore these, right?

Comment (lavalamp) [inline pkg/controller/deployment/util/deployment_util.go]: +1

File: pkg/controller/deployment/util/deployment_util.go (hunk 1 above)
Comment (lavalamp) [inline pkg/controller/deployment/util/deployment_util.go]: I think you can make this a lot simpler.

```go
if len(labels1) > len(labels2) {
  labels1, labels2 = labels2, labels1
}
for k, v := range labels2 {
  if k == extensions.DefaultDeploymentUniqueLabelKey { continue }
  if labels1[k] != v {
    return false, nil
  }
}
```

File: pkg/controller/deployment/util/deployment_util_test.go [hunk 3]
```diff
@@ -462,19 +462,24 @@ func TestEqualIgnoreHash(t *testing.T) {
...
+				t.Errorf("In test case %q%s, expected %v", test.test, reverseString, test.expected)
+			}
+			if t1.Labels == nil || t2.Labels == nil {
+				t.Errorf("In test case %q%s, unexpected labels becomes nil", test.test, reverseString)
+			}
 		}
+		runTest(test.former, test.latter, false)
+		// Test the same case in reverse order
+		runTest(test.latter, test.former, true)
 	}
```
Comment (lavalamp) [inline pkg/controller/deployment/util/deployment_util_test.go]: Add a test that does a deep copy of the input, then compares against the input again after calling equalIgnoreHash, to ensure that equalIgnoreHash doesn't modify its inputs?

Comment (janetkuo) [inline pkg/controller/deployment/util/deployment_util_test.go]: Done

File: pkg/controller/deployment/util/deployment_util.go (hunk 2 above)
Comment (janetkuo) [inline pkg/controller/deployment/util/deployment_util.go]: Fixed

File: pkg/controller/deployment/util/deployment_util.go (hunk 1 above)
Comment (janetkuo) [inline pkg/controller/deployment/util/deployment_util.go]: Done

**Comment by janetkuo** [comment]
Thanks, all comments addressed, ptal

**Comment by lavalamp** [comment]
Please squash, looks good!

**Comment by janetkuo** [comment]
Code squashed, applying tag

File: pkg/controller/deployment/util/deployment_util.go [hunk 4]
```diff
@@ -402,20 +402,34 @@ func ListPods(deployment *extensions.Deployment, getPodList podListFunc) (*api.P
...
+	}
+
+	// Then, compare the templates without comparing their labels
+	template1.Labels, template2.Labels = nil, nil
 	result := api.Semantic.DeepEqual(template1, template2)
-	template1.Labels[extensions.DefaultDeploymentUniqueLabelKey] = hash1
 	return result, nil
 }
 
+func containsHash(labels map[string]string) bool {
```
Comment (AdoHe) [inline pkg/controller/deployment/util/deployment_util.go]: where this function get called?

Comment (janetkuo) [inline pkg/controller/deployment/util/deployment_util.go]: Oops, it's called in the original commit but we don't need it anymore

Comment (janetkuo) [inline pkg/controller/deployment/util/deployment_util.go]: Removed

**Comment by fabioy** [comment]
@janetkuo Cherrypick approved. Please create a cherrypick PR. Thanks.

**Comment by fabioy** [comment]
@janetkuo @lavalamp Ping. Could someone create a cherrypick for this? I'd like to take it for 1.3.5. Thanks.

**Comment by janetkuo** [comment]
@fabioy sorry for the late reply I'll create it

**Comment by janetkuo** [comment]
Cherrypick created #30297

**Pull request merged**